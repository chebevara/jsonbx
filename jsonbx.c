#include "postgres.h"

#include "fmgr.h"
#include "utils/builtins.h"
#include "utils/json.h"
#include "utils/jsonb.h"

#include "jsonbx.h"

PG_MODULE_MAGIC;

PG_FUNCTION_INFO_V1(jsonb_print);

Datum jsonb_print(PG_FUNCTION_ARGS);
char *JsonbToCStringExtended(StringInfo out, char *in, int estimated_len, JsonbOutputKind kind);
static void printCR(StringInfo out, JsonbOutputKind kind);
static void printIndent(StringInfo out, JsonbOutputKind kind, int level);
static void jsonb_put_escaped_value(StringInfo out, JsonbValue * scalarVal);

Datum
jsonb_print(PG_FUNCTION_ARGS)
{
    Jsonb           *jb = PG_GETARG_JSONB(0);
    text            *out;
    int             flags = 0;
    StringInfo      str = makeStringInfo();

    if (PG_GETARG_BOOL(1))
        flags |= PrettyPrint;

    appendBinaryStringInfo(str, "    ", 4); /* VARHDRSZ */
    JsonbToCStringExtended(str, VARDATA(jb), VARSIZE(jb), flags);

    out = (text*)str->data;
    SET_VARSIZE(out, str->len);

    PG_RETURN_TEXT_P(out);
}


/*
 * JsonbToCStringExtended
 *       Converts jsonb value to a C-string (take JsonbOutputKind into account).
 * See the original function JsonbToCString in the jsonb.c
 */
char *
JsonbToCStringExtended(StringInfo out, JsonbSuperHeader in, int estimated_len, JsonbOutputKind kind)
{
    bool           first = true;
    JsonbIterator *it;
    int            type = 0;
    JsonbValue     v;
    int            level = 0;
    bool           redo_switch = false;

    if (out == NULL)
        out = makeStringInfo();

    enlargeStringInfo(out, (estimated_len >= 0) ? estimated_len : 64);

    it = JsonbIteratorInit(in);

    while (redo_switch ||
           ((type = JsonbIteratorNext(&it, &v, false)) != WJB_DONE))
    {
        redo_switch = false;
        switch (type)
        {
            case WJB_BEGIN_ARRAY:
                if (!first)
                    appendBinaryStringInfo(out, ", ", 2);
                first = true;

                if (!v.val.array.rawScalar)
                    printCR(out, kind);
                    printIndent(out, kind, level);

                    appendStringInfoChar(out, '[');
                level++;

                printCR(out, kind);
                printIndent(out, kind, level);

                break;
            case WJB_BEGIN_OBJECT:
                if (!first)
                    appendBinaryStringInfo(out, ", ", 2);
                first = true;

                printCR(out, kind);
                printIndent(out, kind, level);

                appendStringInfoCharMacro(out, '{');

                level++;
                break;
            case WJB_KEY:
                if (!first)
                    appendBinaryStringInfo(out, ", ", 2);
                first = true;

                printCR(out, kind);
                printIndent(out, kind, level);

                /* json rules guarantee this is a string */
                jsonb_put_escaped_value(out, &v);
                appendBinaryStringInfo(out, ": ", 2);

                type = JsonbIteratorNext(&it, &v, false);
                if (type == WJB_VALUE)
                {
                    first = false;
                    jsonb_put_escaped_value(out, &v);
                }
                else
                {
                    Assert(type == WJB_BEGIN_OBJECT || type == WJB_BEGIN_ARRAY);

                    /*
                     * We need to rerun the current switch() since we need to
                     * output the object which we just got from the iterator
                     * before calling the iterator again.
                     */
                    redo_switch = true;
                }
                break;
            case WJB_ELEM:
                if (!first) {
                    appendBinaryStringInfo(out, ", ", 2);

                    printCR(out, kind);
                    printIndent(out, kind, level);
                }
                else {
                    first = false;
                }

                jsonb_put_escaped_value(out, &v);
                break;
            case WJB_END_ARRAY:
                level--;

                printCR(out, kind);
                printIndent(out, kind, level);

                if (!v.val.array.rawScalar)
                    appendStringInfoChar(out, ']');
                first = false;
                break;
            case WJB_END_OBJECT:
                level--;

                printCR(out, kind);
                printIndent(out, kind, level);

                appendStringInfoCharMacro(out, '}');
                first = false;
                break;
            default:
                elog(ERROR, "unknown flag of jsonb iterator");
        }
    }

    Assert(level == 0);

    return out->data;
}


static void
printCR(StringInfo out, JsonbOutputKind kind)
{
    if (kind & PrettyPrint)
        appendBinaryStringInfo(out, "    ", 4);
        appendStringInfoCharMacro(out, '\n');
}

static void
printIndent(StringInfo out, JsonbOutputKind kind, int level)
{
    if (kind & PrettyPrint)
    {
        int i;
        for(i=0; i<4*level; i++)
            appendStringInfoCharMacro(out, ' ');
    }
}


static void
jsonb_put_escaped_value(StringInfo out, JsonbValue * scalarVal)
{
    switch (scalarVal->type)
    {
        case jbvNull:
            appendBinaryStringInfo(out, "null", 4);
            break;
        case jbvString:
            escape_json(out, pnstrdup(scalarVal->val.string.val, scalarVal->val.string.len));
            break;
        case jbvNumeric:
            appendStringInfoString(out,
                                   DatumGetCString(DirectFunctionCall1(numeric_out,
                                                                       PointerGetDatum(scalarVal->val.numeric))));
            break;
        case jbvBool:
            if (scalarVal->val.boolean)
                appendBinaryStringInfo(out, "true", 4);
            else
                appendBinaryStringInfo(out, "false", 5);
            break;
        default:
            elog(ERROR, "unknown jsonb scalar type");
    }
}
