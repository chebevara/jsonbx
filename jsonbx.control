# test extension
comment = 'Jsonb extension'
default_version = '1.0'
module_pathname = '$libdir/jsonbx'
relocatable = true
