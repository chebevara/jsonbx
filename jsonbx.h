#ifndef __JSONBX_H__
#define __JSONBX_H__


typedef enum JsonbOutputKind {
    PrettyPrint = 0x10
} JsonbOutputKind;


#endif
