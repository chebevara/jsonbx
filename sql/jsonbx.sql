CREATE EXTENSION jsonbx;

select jsonb_print('{"a": "test", "b": [1, 2, 3], "c": "test3", "d":{"dd": "test4", "dd2":{"ddd": "test5"}}}'::jsonb, true);

select jsonb_concat('{"d": "test", "a": [1, 2]}'::jsonb, '{"g": "test2", "c": {"c1":1, "c2":2}}'::jsonb);

select '{"aa":1 , "b":2, "cq":3}'::jsonb || '{"cq":"l", "b":"g", "fg":false}';
select '{"aa":1 , "b":2, "cq":3}'::jsonb || '{"aq":"l"}';
select '{"aa":1 , "b":2, "cq":3}'::jsonb || '{"aa":"l"}';
select '{"aa":1 , "b":2, "cq":3}'::jsonb || '{}';
select '{}'::jsonb || '{"cq":"l", "b":"g", "fg":false}';
select pg_column_size('{}'::jsonb || '{}'::jsonb) = pg_column_size('{}'::jsonb);
select pg_column_size('{"aa":1}'::jsonb || '{"b":2}'::jsonb) = pg_column_size('{"aa":1, "b":2}'::jsonb);
select pg_column_size('{"aa":1, "b":2}'::jsonb || '{}'::jsonb) = pg_column_size('{"aa":1, "b":2}'::jsonb);
select pg_column_size('{}'::jsonb || '{"aa":1, "b":2}'::jsonb) = pg_column_size('{"aa":1, "b":2}'::jsonb);

select jsonb_delete('{"a":1 , "b":2, "c":3}'::jsonb, 'a');
select jsonb_delete('{"a":null , "b":2, "c":3}'::jsonb, 'a');
select jsonb_delete('{"a":1 , "b":2, "c":3}'::jsonb, 'b');
select jsonb_delete('{"a":1 , "b":2, "c":3}'::jsonb, 'c');
select jsonb_delete('{"a":1 , "b":2, "c":3}'::jsonb, 'd');
select '{"a":1 , "b":2, "c":3}'::jsonb - 'a'::text;
select '{"a":null , "b":2, "c":3}'::jsonb - 'a'::text;
select '{"a":1 , "b":2, "c":3}'::jsonb - 'b'::text;
select '{"a":1 , "b":2, "c":3}'::jsonb - 'c'::text;
select '{"a":1 , "b":2, "c":3}'::jsonb - 'd'::text;
select pg_column_size('{"a":1 , "b":2, "c":3}'::jsonb - 'b'::text) = pg_column_size('{"a":1, "b":2}'::jsonb);

select '["a","b","c"]'::jsonb - 3;
select '["a","b","c"]'::jsonb - 2;
select '["a","b","c"]'::jsonb - 1;
select '["a","b","c"]'::jsonb - 0;
select '["a","b","c"]'::jsonb - -1;
select '["a","b","c"]'::jsonb - -2;
select '["a","b","c"]'::jsonb - -3;
select '["a","b","c"]'::jsonb - -4;

select '{"a":1, "b":2, "c":3}'::jsonb - 3;
select '{"a":1, "b":2, "c":3}'::jsonb - 2;
select '{"a":1, "b":2, "c":3}'::jsonb - 1;
select '{"a":1, "b":2, "c":3}'::jsonb - 0;
select '{"a":1, "b":2, "c":3}'::jsonb - -1;
select '{"a":1, "b":2, "c":3}'::jsonb - -2;
select '{"a":1, "b":2, "c":3}'::jsonb - -3;
select '{"a":1, "b":2, "c":3}'::jsonb - -4;

select jsonb_replace('{"n":null, "a":1, "b":[1,2], "c":{"1":2}, "d":{"1":[2,3]}}'::jsonb, '{n}', '[1,2,3]');
select jsonb_replace('{"n":null, "a":1, "b":[1,2], "c":{"1":2}, "d":{"1":[2,3]}}'::jsonb, '{b,-1}', '[1,2,3]');
select jsonb_replace('{"n":null, "a":1, "b":[1,2], "c":{"1":2}, "d":{"1":[2,3]}}'::jsonb, '{d,1,0}', '[1,2,3]');
select jsonb_replace('{"n":null, "a":1, "b":[1,2], "c":{"1":2}, "d":{"1":[2,3]}}'::jsonb, '{d,NULL,0}', '[1,2,3]');

select jsonb_replace('{"n":null, "a":1, "b":[1,2], "c":{"1":2}, "d":{"1":[2,3]}}'::jsonb, '{n}', '{"1": 2}');
select jsonb_replace('{"n":null, "a":1, "b":[1,2], "c":{"1":2}, "d":{"1":[2,3]}}'::jsonb, '{b,-1}', '{"1": 2}');
select jsonb_replace('{"n":null, "a":1, "b":[1,2], "c":{"1":2}, "d":{"1":[2,3]}}'::jsonb, '{d,1,0}', '{"1": 2}');
select jsonb_replace('{"n":null, "a":1, "b":[1,2], "c":{"1":2}, "d":{"1":[2,3]}}'::jsonb, '{d,NULL,0}', '{"1": 2}');
