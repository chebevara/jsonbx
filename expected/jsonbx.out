CREATE EXTENSION jsonbx;
select jsonb_print('{"a": "test", "b": [1, 2, 3], "c": "test3", "d":{"dd": "test4", "dd2":{"ddd": "test5"}}}'::jsonb, true);
          jsonb_print           
--------------------------------
                               +
 {                             +
     "a": "test",              +
     "b":                      +
     [                         +
         1,                    +
         2,                    +
         3                     +
     ],                        +
     "c": "test3",             +
     "d":                      +
     {                         +
         "dd": "test4",        +
         "dd2":                +
         {                     +
             "ddd": "test5"    +
         }                     +
     }                         +
 }
(1 row)

select jsonb_concat('{"d": "test", "a": [1, 2]}'::jsonb, '{"g": "test2", "c": {"c1":1, "c2":2}}'::jsonb);
                           jsonb_concat                            
-------------------------------------------------------------------
 {"a": [1, 2], "c": {"c1": 1, "c2": 2}, "d": "test", "g": "test2"}
(1 row)

select '{"aa":1 , "b":2, "cq":3}'::jsonb || '{"cq":"l", "b":"g", "fg":false}';
                  ?column?                   
---------------------------------------------
 {"b": "g", "aa": 1, "cq": "l", "fg": false}
(1 row)

select '{"aa":1 , "b":2, "cq":3}'::jsonb || '{"aq":"l"}';
               ?column?                
---------------------------------------
 {"b": 2, "aa": 1, "aq": "l", "cq": 3}
(1 row)

select '{"aa":1 , "b":2, "cq":3}'::jsonb || '{"aa":"l"}';
           ?column?           
------------------------------
 {"b": 2, "aa": "l", "cq": 3}
(1 row)

select '{"aa":1 , "b":2, "cq":3}'::jsonb || '{}';
          ?column?          
----------------------------
 {"b": 2, "aa": 1, "cq": 3}
(1 row)

select '["a", "b"]'::jsonb || '["c"]';
    ?column?     
-----------------
 ["a", "b", "c"]
(1 row)

select '["a", "b"]'::jsonb || '"c"';
    ?column?     
-----------------
 ["a", "b", "c"]
(1 row)

select '"a"'::jsonb || '{"a":1}';
ERROR:  invalid concatnation of jsonb objects
select '["a"]'::jsonb || '{"a":1}';
ERROR:  invalid concatnation of jsonb objects
select '{"a": 1}'::jsonb || '["b"]';
ERROR:  invalid concatnation of jsonb objects
select '{}'::jsonb || '{"cq":"l", "b":"g", "fg":false}';
              ?column?              
------------------------------------
 {"b": "g", "cq": "l", "fg": false}
(1 row)

select pg_column_size('{}'::jsonb || '{}'::jsonb) = pg_column_size('{}'::jsonb);
 ?column? 
----------
 t
(1 row)

select pg_column_size('{"aa":1}'::jsonb || '{"b":2}'::jsonb) = pg_column_size('{"aa":1, "b":2}'::jsonb);
 ?column? 
----------
 t
(1 row)

select pg_column_size('{"aa":1, "b":2}'::jsonb || '{}'::jsonb) = pg_column_size('{"aa":1, "b":2}'::jsonb);
 ?column? 
----------
 t
(1 row)

select pg_column_size('{}'::jsonb || '{"aa":1, "b":2}'::jsonb) = pg_column_size('{"aa":1, "b":2}'::jsonb);
 ?column? 
----------
 t
(1 row)

select jsonb_delete('{"a":1 , "b":2, "c":3}'::jsonb, 'a');
   jsonb_delete   
------------------
 {"b": 2, "c": 3}
(1 row)

select jsonb_delete('{"a":null , "b":2, "c":3}'::jsonb, 'a');
   jsonb_delete   
------------------
 {"b": 2, "c": 3}
(1 row)

select jsonb_delete('{"a":1 , "b":2, "c":3}'::jsonb, 'b');
   jsonb_delete   
------------------
 {"a": 1, "c": 3}
(1 row)

select jsonb_delete('{"a":1 , "b":2, "c":3}'::jsonb, 'c');
   jsonb_delete   
------------------
 {"a": 1, "b": 2}
(1 row)

select jsonb_delete('{"a":1 , "b":2, "c":3}'::jsonb, 'd');
       jsonb_delete       
--------------------------
 {"a": 1, "b": 2, "c": 3}
(1 row)

select '{"a":1 , "b":2, "c":3}'::jsonb - 'a'::text;
     ?column?     
------------------
 {"b": 2, "c": 3}
(1 row)

select '{"a":null , "b":2, "c":3}'::jsonb - 'a'::text;
     ?column?     
------------------
 {"b": 2, "c": 3}
(1 row)

select '{"a":1 , "b":2, "c":3}'::jsonb - 'b'::text;
     ?column?     
------------------
 {"a": 1, "c": 3}
(1 row)

select '{"a":1 , "b":2, "c":3}'::jsonb - 'c'::text;
     ?column?     
------------------
 {"a": 1, "b": 2}
(1 row)

select '{"a":1 , "b":2, "c":3}'::jsonb - 'd'::text;
         ?column?         
--------------------------
 {"a": 1, "b": 2, "c": 3}
(1 row)

select pg_column_size('{"a":1 , "b":2, "c":3}'::jsonb - 'b'::text) = pg_column_size('{"a":1, "b":2}'::jsonb);
 ?column? 
----------
 t
(1 row)

select '["a","b","c"]'::jsonb - 3;
    ?column?     
-----------------
 ["a", "b", "c"]
(1 row)

select '["a","b","c"]'::jsonb - 2;
  ?column?  
------------
 ["a", "b"]
(1 row)

select '["a","b","c"]'::jsonb - 1;
  ?column?  
------------
 ["a", "c"]
(1 row)

select '["a","b","c"]'::jsonb - 0;
  ?column?  
------------
 ["b", "c"]
(1 row)

select '["a","b","c"]'::jsonb - -1;
  ?column?  
------------
 ["a", "b"]
(1 row)

select '["a","b","c"]'::jsonb - -2;
  ?column?  
------------
 ["a", "c"]
(1 row)

select '["a","b","c"]'::jsonb - -3;
  ?column?  
------------
 ["b", "c"]
(1 row)

select '["a","b","c"]'::jsonb - -4;
    ?column?     
-----------------
 ["a", "b", "c"]
(1 row)

select '{"a":1, "b":2, "c":3}'::jsonb - 3;
         ?column?         
--------------------------
 {"a": 1, "b": 2, "c": 3}
(1 row)

select '{"a":1, "b":2, "c":3}'::jsonb - 2;
     ?column?     
------------------
 {"a": 1, "b": 2}
(1 row)

select '{"a":1, "b":2, "c":3}'::jsonb - 1;
     ?column?     
------------------
 {"a": 1, "c": 3}
(1 row)

select '{"a":1, "b":2, "c":3}'::jsonb - 0;
     ?column?     
------------------
 {"b": 2, "c": 3}
(1 row)

select '{"a":1, "b":2, "c":3}'::jsonb - -1;
     ?column?     
------------------
 {"a": 1, "b": 2}
(1 row)

select '{"a":1, "b":2, "c":3}'::jsonb - -2;
     ?column?     
------------------
 {"a": 1, "c": 3}
(1 row)

select '{"a":1, "b":2, "c":3}'::jsonb - -3;
     ?column?     
------------------
 {"b": 2, "c": 3}
(1 row)

select '{"a":1, "b":2, "c":3}'::jsonb - -4;
         ?column?         
--------------------------
 {"a": 1, "b": 2, "c": 3}
(1 row)

select jsonb_replace('{"n":null, "a":1, "b":[1,2], "c":{"1":2}, "d":{"1":[2,3]}}'::jsonb, '{n}', '[1,2,3]');
                              jsonb_replace                               
--------------------------------------------------------------------------
 {"a": 1, "b": [1, 2], "c": {"1": 2}, "d": {"1": [2, 3]}, "n": [1, 2, 3]}
(1 row)

select jsonb_replace('{"n":null, "a":1, "b":[1,2], "c":{"1":2}, "d":{"1":[2,3]}}'::jsonb, '{b,-1}', '[1,2,3]');
                                jsonb_replace                                
-----------------------------------------------------------------------------
 {"a": 1, "b": [1, [1, 2, 3]], "c": {"1": 2}, "d": {"1": [2, 3]}, "n": null}
(1 row)

select jsonb_replace('{"n":null, "a":1, "b":[1,2], "c":{"1":2}, "d":{"1":[2,3]}}'::jsonb, '{d,1,0}', '[1,2,3]');
                                jsonb_replace                                
-----------------------------------------------------------------------------
 {"a": 1, "b": [1, 2], "c": {"1": 2}, "d": {"1": [[1, 2, 3], 3]}, "n": null}
(1 row)

select jsonb_replace('{"n":null, "a":1, "b":[1,2], "c":{"1":2}, "d":{"1":[2,3]}}'::jsonb, '{d,NULL,0}', '[1,2,3]');
                            jsonb_replace                            
---------------------------------------------------------------------
 {"a": 1, "b": [1, 2], "c": {"1": 2}, "d": {"1": [2, 3]}, "n": null}
(1 row)

select jsonb_replace('{"n":null, "a":1, "b":[1,2], "c":{"1":2}, "d":{"1":[2,3]}}'::jsonb, '{n}', '{"1": 2}');
                              jsonb_replace                              
-------------------------------------------------------------------------
 {"a": 1, "b": [1, 2], "c": {"1": 2}, "d": {"1": [2, 3]}, "n": {"1": 2}}
(1 row)

select jsonb_replace('{"n":null, "a":1, "b":[1,2], "c":{"1":2}, "d":{"1":[2,3]}}'::jsonb, '{b,-1}', '{"1": 2}');
                               jsonb_replace                                
----------------------------------------------------------------------------
 {"a": 1, "b": [1, {"1": 2}], "c": {"1": 2}, "d": {"1": [2, 3]}, "n": null}
(1 row)

select jsonb_replace('{"n":null, "a":1, "b":[1,2], "c":{"1":2}, "d":{"1":[2,3]}}'::jsonb, '{d,1,0}', '{"1": 2}');
                               jsonb_replace                                
----------------------------------------------------------------------------
 {"a": 1, "b": [1, 2], "c": {"1": 2}, "d": {"1": [{"1": 2}, 3]}, "n": null}
(1 row)

select jsonb_replace('{"n":null, "a":1, "b":[1,2], "c":{"1":2}, "d":{"1":[2,3]}}'::jsonb, '{d,NULL,0}', '{"1": 2}');
                            jsonb_replace                            
---------------------------------------------------------------------
 {"a": 1, "b": [1, 2], "c": {"1": 2}, "d": {"1": [2, 3]}, "n": null}
(1 row)

-- we can't use simple text for the last parameter like this:
--
-- =# select jsonb_replace('{"n":null, "a":1, "b":[1,2], "c":{"1":2}, "d":{"1":[2,3]}}'::jsonb, '{b,-1}', '"test"');
--
-- because it would be interpreted as an array (so we can't identify, is it a text or an array?) and this lead to
--
-- {"a": 1, "b": [1, "test", "c": {"1": 2}, "d": {"1": [2, 3]}, "n": null}
select jsonb_replace('{"n":null, "a":1, "b":[1,2], "c":{"1":2}, "d":{"1":[2,3]}}'::jsonb, '{b,-1}', '{"f": "test"}');
                                  jsonb_replace                                  
---------------------------------------------------------------------------------
 {"a": 1, "b": [1, {"f": "test"}], "c": {"1": 2}, "d": {"1": [2, 3]}, "n": null}
(1 row)

